stream/stream.go

```diff
@@ -4,229 +4,230 @@ package stream
-type Pubsub interface {
-	Publish(context.Context, string, []byte) error
-	Subscribe(context.Context, string) chan []byte
-}
-
-type StreamConfig struct {
-	Logger          log.Logger
-	ID              string
-	PubsubTopic     string // pubsub topic name for block submissions
-	StreamQueueSize int
+type PubSub interface {
+	Publish(context.Context, transport.Message) error
+	Subscribe(context.Context) transport.Subscription
 }
 
 type State interface {
 	ForkVersion(epoch structs.Slot) structs.ForkVersion
 	HeadSlot() structs.Slot
 }
 
+type Metrics interface {
+	Register(prometheus.Collector) error
+}
+
 type Client struct {
-	Pubsub Pubsub
+	once sync.Once
 
-	builderBidIn     chan []byte
-	builderBidOut    chan structs.BuilderBidExtended
-	cacheIn          chan []byte
-	cacheOut         chan structs.BlockAndTraceExtended
-	slotDeliveredIn  chan []byte
-	slotDeliveredOut chan uint64
+	Logger                log.Logger
+	State                 State
+	Bids, Cache           PubSub
+	QueueSize, NumWorkers int
 
-	Config StreamConfig
-	Logger log.Logger
+	Metrics Metrics
+	m       streamMetrics
 
-	m StreamMetrics
+	ctx    context.Context
+	cancel context.CancelFunc
 
-	//slotDelivered chan structs.Slot
+	builderBidOut chan structs.BuilderBidExtended
+	cacheOut      chan structs.BlockAndTraceExtended
 
 	st State
 }
 
-func NewClient(ps Pubsub, st State, cfg StreamConfig) *Client {
-	s := Client{
-		Pubsub: ps,
-		st:     st,
-
-		builderBidIn:     make(chan []byte, cfg.StreamQueueSize),
-		builderBidOut:    make(chan structs.BuilderBidExtended, cfg.StreamQueueSize),
-		cacheIn:          make(chan []byte, cfg.StreamQueueSize),
-		cacheOut:         make(chan structs.BlockAndTraceExtended, cfg.StreamQueueSize),
-		slotDeliveredIn:  make(chan []byte, cfg.StreamQueueSize),
-		slotDeliveredOut: make(chan uint64, cfg.StreamQueueSize),
+func (c *Client) init() {
+	c.once.Do(func() {
+		c.initMetrics()
 
-		Config: cfg,
-		Logger: cfg.Logger.WithField("subService", "stream").WithField("type", "redis"),
-	}
+		if c.Logger == nil {
+			c.Logger = log.New()
+		}
 
-	s.initMetrics()
+		c.builderBidOut = make(chan structs.BuilderBidExtended, c.QueueSize)
+		c.cacheOut = make(chan structs.BlockAndTraceExtended, c.QueueSize)
 
-	return &s
+		c.ctx, c.cancel = context.WithCancel(context.Background())
+		go c.subscribe(c.Bids, c.handleBid)
+		go c.subscribe(c.Cache, c.handleCache)
+	})
 }
 
-func (s *Client) RunSubscriberParallel(ctx context.Context, num uint) error {
-	s.builderBidIn = s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic+BidTopic)
-	s.cacheIn = s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic+CacheTopic)
-	s.slotDeliveredIn = s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic+SlotDeliveredTopic)
-
-	for i := uint(0); i < num; i++ {
-		go s.RunBlockCacheSubscriber(ctx)
-		go s.RunBuilderBidSubscriber(ctx)
-	}
-
-	go s.RunSlotDeliveredSubscriber(ctx)
-
+func (c *Client) Close() error {
+	c.once.Do(func() {
+		c.ctx, c.cancel = context.WithCancel(context.Background())
```

Comment:
this may not work if you overwrite c.ctx and close it - for something that was already inherited the previous context downwards

Reply:
`sync.Once` will only ever fire once.  It either sets the context here, or in the `init` function.  This is intended to prevent panics due to either (1) nil `c.cancel` in this function or (2) a nil context in a concurrent call to `init`.

Reply:
e1ee8bef260ce049e95b68f0d131eb65eb10a5f8